vdr/didx509/resolver.go

```diff
@@ -0,0 +1,232 @@
+	ErrWrongCN                 = errors.New("query does not match the subject CN")
+	ErrWrongLocality           = errors.New("query does not match the subject Locality")
+	ErrWrongOrganization       = errors.New("query does not match the subject Organization")
+	ErrWrongOrganizationalUnit = errors.New("query does not match the subject OrganizationalUnit")
+	ErrWrongSanOtherName       = errors.New("the SAN otherName does not match the query")
+	ErrWrongSanDns             = errors.New("the SAN DNS does not match the query")
+	ErrWrongSanEmailAddresses  = errors.New("the SAN EmailAddresses does not match the query")
+	ErrWrongSanIPAddresses     = errors.New("the SAN IPAddresses does not match the query")
+	ErrDidMalformed            = errors.New("did:x509 is malformed")
+	ErrDidVersion              = errors.New("did:x509 does not have version 0")
+	ErrDidSanMalformed         = errors.New("did:x509 policy is malformed")
+	ErrX509ChainMissing        = errors.New("x509 rootCert chain is missing")
+)
+
+var _ resolver.DIDResolver = &Resolver{}
+
+// NewResolver creates a new Resolver.
+func NewResolver(pkiValidator pki.Validator) *Resolver {
+	return &Resolver{
+		pkiValidator: pkiValidator,
+	}
+}
+
+type Resolver struct {
+	pkiValidator pki.Validator
+}
+
+type X509DidReference struct {
+	Method      string
+	RootCertRef string
+	PolicyName  string
+	PolicyValue string
+}
+
+// Resolve resolves a DID document given its identifier and corresponding metadata.
+func (r Resolver) Resolve(id did.DID, metadata *resolver.ResolveMetadata) (*did.Document, *resolver.DocumentMetadata, error) {
+	if id.Method != MethodName {
+		return nil, nil, fmt.Errorf("unsupported DID method: %s", id.Method)
+	}
+	ref, err := parseX509Did(id)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if metadata.X509CertChain == nil {
+		return nil, nil, ErrX509ChainMissing
+	}
+	chain, err := parseChain(metadata)
+	if err != nil {
+		return nil, nil, err
+	}
+	_, err = findCertificateByHash(chain, ref.RootCertRef, ref.Method)
+	if err != nil {
+		return nil, nil, err
+	}
+	validationCert, err := findCertificateByHash(chain, metadata.X509CertThumbprint, "sha1")
+	if err != nil {
+		validationCert, err = findCertificateByHash(chain, metadata.X509CertThumbprintS256, "sha256")
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	err = validatePolicy(ref, validationCert)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = r.pkiValidator.Validate(chain)
+	if err != nil {
+		return nil, nil, err
+	}
+	document, err := createDidDocument(id, validationCert)
+	if err != nil {
+		return nil, nil, err
+	}
+	return document, &resolver.DocumentMetadata{}, err
+}
+
+// validatePolicy validates a certificate against a given X509DidReference and its policy.
+func validatePolicy(ref *X509DidReference, cert *x509.Certificate) error {
+
+	switch ref.PolicyName {
+	case "subject":
+		keyValue := strings.Split(ref.PolicyValue, ":")
+		if len(keyValue)%2 != 0 {
+			return ErrDidSanMalformed
```

Comment:
I think this should've been subject malformed?

Reply:
Correct. Fixed.